### PR TITLE
Lock merge migration script

### PR DESCRIPTION
**Pull request by mconcat**
Closes: #867 

- [X] Lockup jitter merge
- [ ] Accumulation store reset

## Description

______

For contributor use:

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/gaia/blob/main/CONTRIBUTING.md#pr-targeting))
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer

### CODE REVIEW

File: x/lockup/keeper/migration.go [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+	HourDuration, _    = time.ParseDuration("1h")
+	DayDuration, _     = time.ParseDuration("24h")
+	WeekDuration, _    = time.ParseDuration("168h")
+	TwoWeekDuration, _ = time.ParseDuration("336h")
+	BaselineDurations  = []time.Duration{DayDuration, WeekDuration, TwoWeekDuration}
+)
+
+func normalizeDuration(baselineDurations []time.Duration, allowedDiff time.Duration, duration time.Duration) time.Duration {
+	for _, base := range baselineDurations {
+		if base <= duration && duration < base+allowedDiff {
```
Comment (ValarDragon) [inline x/lockup/keeper/migration.go:21]:
```suggestion
// baselineDurations is expected to be sorted by the caller 
func normalizeDuration(baselineDurations []time.Duration, allowedDiff time.Duration, duration time.Duration) time.Duration {
	for _, base := range baselineDurations {
	// if base > duration, continue to next base size.
	// if base <= duration, then we are in a duration that is greater than or equal to base size.
	// If its within base + allowed diff, we set it to base.
		if base <= duration && duration < base+allowedDiff {
```

File: x/lockup/keeper/migration.go [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+		normals := make(map[string]uint64)
+		for _, lock := range k.GetAccountPeriodLocks(ctx, addr) {
+			// ignore multilocks
+			if len(lock.Coins) > 1 {
+				continue
+			}
+			// ignore unlocking locks; they will be removed from the state anyway
+			if lock.IsUnlocking() {
+				continue
+			}
```
Comment (ValarDragon) [inline x/lockup/keeper/migration.go:47]: Nice!

File: x/lockup/keeper/migration.go [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+			coin := lock.Coins[0]
+			normalizedDuration := normalizeDuration(baselineDurations, durationDiff, lock.Duration)
+			key := addr.String() + "/" + coin.Denom + "/" + strconv.FormatInt(int64(normalizedDuration), 10)
```
Comment (ValarDragon) [inline x/lockup/keeper/migration.go:50]: What is this key? Is there a standard function for it?

Comment (mconcat) [reply x/lockup/keeper/migration.go:50]: key is just a triplet for (address, denom, duration). We can do
```suggestion
			key := strings.Join([]string{addr.String(), coin.Denom, strconv.FormatInt(int64(normalizedDuration), 10)}, "/")
```

Comment (ValarDragon) [reply x/lockup/keeper/migration.go:53]: Oh its a key just for all the normalized locks, we don't need to parse it ever / it doesn't affect state.

Comment (ValarDragon) [reply x/lockup/keeper/migration.go:53]: SGTM! Latter SGTM

**Comment by mconcat** [comment]
Having a weird build error that is not reproducible on local

File: x/lockup/keeper/migration.go [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+	// if base <= duration, then we are in a duration that is greater than or equal to base size.
+	// If its within base + allowed diff, we set it to base.
+		if base <= duration && duration < base+allowedDiff {
+			return base
+		}
+	}
+	return duration
+}
+
+// between duration windows of baselineDurations.map((duration) => (duration, duration+durationDiff)).
```
Comment (ValarDragon) [inline x/lockup/keeper/migration.go:31]:
```suggestion
// we combine all locks between duration windows of baselineDurations.map((duration) => (duration, duration+durationDiff)) into a single new lock.
// This means every lock ID will get updated.
```

**Review by ValarDragon — APPROVED** [review]
Nice job on this! it looks solid to me!

**Pull request merged**